readme.md
## [![npm][npmjs-img]][npmjs-url] [![mit license][license-img]][license-url] [![build status][travis-img]][travis-url] [![coverage status][coveralls-img]][coveralls-url] [![deps status][daviddm-img]][daviddm-url]

> Get current week number of the current year,given Date format string or Date object.

## Install
```
npm i --save current-week-number
npm test
current-week-number --help
```


## API
> For more use-cases see the [tests](./test.js)

### [currentWeekNumber](./index.js#L40)
> Get week number of the current date/year or given valid `Date` string format

- `[date]` **{String}** every valid Date-ish string format
- `return` **{Number}**

**Example:**

```js
var currentWeekNumber = require('current-week-number');

// june 27, 2014
currentWeekNumber();
//=> 26

currentWeekNumber('March 24, 2015');
//=> 13

currentWeekNumber(new Date('March 24, 2015'));
//=> 13

currentWeekNumber('03/24/2016');
//=> 12, cuz' year is leap

currentWeekNumber('August 07, 2015');
//=> 32

currentWeekNumber(new Date('August 07, 2016'));
//=> 31

currentWeekNumber('02/16/2015');
//=> 8

currentWeekNumber('September 15, 2126');
//=> 37

currentWeekNumber('02/17/2012');
//=> 7
```


## CLI
> You can just run `current-week-number --help` for more information


## Author
**Charlike Mike Reagent**
+ [gratipay/tunnckoCore][author-gratipay]
+ [twitter/tunnckoCore][author-twitter]
+ [github/tunnckoCore][author-github]
+ [npmjs/tunnckoCore][author-npmjs]
+ [more ...][contrib-more]


## License [![MIT license][license-img]][license-url]
Copyright (c) 2014-2015 [Charlike Mike Reagent][contrib-more], [contributors][contrib-graf].  
Released under the [`MIT`][license-url] license.


[npmjs-url]: http://npm.im/current-week-number
[npmjs-img]: https://img.shields.io/npm/v/current-week-number.svg?style=flat&label=current-week-number

[coveralls-url]: https://coveralls.io/r/datetime/current-week-number?branch=master
[coveralls-img]: https://img.shields.io/coveralls/datetime/current-week-number.svg?style=flat

[license-url]: https://github.com/datetime/current-week-number/blob/master/license.md
[license-img]: https://img.shields.io/badge/license-MIT-blue.svg?style=flat

[travis-url]: https://travis-ci.org/datetime/current-week-number
[travis-img]: https://img.shields.io/travis/datetime/current-week-number.svg?style=flat

[daviddm-url]: https://david-dm.org/datetime/current-week-number
[daviddm-img]: https://img.shields.io/david/dev/datetime/current-week-number.svg?style=flat

[author-gratipay]: https://gratipay.com/tunnckoCore
[author-twitter]: https://twitter.com/tunnckoCore
[author-github]: https://github.com/tunnckoCore
[author-npmjs]: https://npmjs.org/~tunnckocore

[contrib-more]: http://j.mp/1stW47C
[contrib-graf]: https://github.com/datetime/current-week-number/graphs/contributors

***

_Powered and automated by [kdf](https://github.com/tunnckoCore), March 02, 2015_

history.md


## v1.0.7 / March 02, 2015
- Release v1.0.7 / npm@v1.0.7
- update david-dm badge
- add cli usage
- fix travis, iojs support
- add more examples
- add cli, ref #1

## v1.0.6 / February 4, 2015
- Release v1.0.6 / npm@v1.0.6
- update description
- update docs
- update tests
- run lint
- remove `mocha-lcov-reporter`
- merge pull request #2 from wankdanker/master 
- remove cli.js reference in package.json
- already accepts and Date object

## v1.0.5 / January 26, 2015
- Release v1.0.5 / npm@v1.0.5
- refactor, update tests, coverage

## v1.0.4 / December 12, 2014
- Release v1.0.4 / npm@v1.0.4
- fixes, tweaks, dotfiles, docs, etc
- nothing special

## v1.0.3 / December 12, 2014
- Release v1.0.3 / npm@v1.0.3
- fix links to [datetime org][datetime-url]

## v1.0.2 / December 12, 2014
- Release v1.0.2 / npm@v1.0.2
- moved to [datetime org][datetime-url]
- remove should from deps, use `assert`
- update tests

## v1.0.1 / December 08, 2014
- Release v1.0.1 / npm@v1.0.1
- remove `moment` as dep

## v1.0.0 / December 08, 2014
- Release v1.0.0 / npm@v1.0.0
- add cli support
- no breaking changes, just init release

## v0.1.1 / December 08, 2014
- Release v0.1.1 / npm@v0.1.1
- dotfiles, refactoring
- update tests
- update devDeps

## v0.1.0 / June 26, 2014
- Release v0.1.0 / npm@v0.1.0
- initial commits


[datetime-url]: https://github.com/datetime

test.js
/**
 * current-week-number <https://github.com/datetime/current-week-number>
 *
 * Copyright (c) 2014-2015 Charlike Mike Reagent, contributors.
 * Released under the MIT license.
 */

'use strict';

var assert = require('assert');
var currentWeekNumber = require('./index');

describe('current-week-number:', function() {
  it('should receive valid date format string', function(done) {
    assert.strictEqual(currentWeekNumber('March 24, 2015'), 13);
    assert.strictEqual(currentWeekNumber('03/24/2016'), 12);
    assert.strictEqual(currentWeekNumber('August 07, 2016'), 31);
    assert.strictEqual(currentWeekNumber('August 07, 2015'), 32);
    assert.strictEqual(currentWeekNumber('02/16/2015'), 8);
    assert.strictEqual(currentWeekNumber('12/15/2014'), 51);
    done();
  });
  it('should receive date object', function(done) {
    assert.strictEqual(currentWeekNumber(new Date('March 24, 2015')), 13);
    assert.strictEqual(currentWeekNumber(new Date('August 07, 2016')), 31);
    done();
  });
  it('should get current week number when empty string format', function(done) {
    assert.strictEqual(typeof currentWeekNumber(''), 'number');
    done();
  });
  it('should get current week number, if no valid Date format or Date Object', function(done) {
    assert.strictEqual(typeof currentWeekNumber(/regex/g), 'number');
    assert.strictEqual(typeof currentWeekNumber(true), 'number');
    assert.strictEqual(typeof currentWeekNumber(false), 'number');
    assert.strictEqual(typeof currentWeekNumber(222), 'number');
    assert.strictEqual(typeof currentWeekNumber({obj: 123}), 'number');
    assert.strictEqual(typeof currentWeekNumber([1,2,3]), 'number');
    assert.strictEqual(typeof currentWeekNumber(), 'number');
    done();
  });
});


cli.js
#!/usr/bin/env node
/**
 * current-week-number <https://github.com/datetime/current-week-number>
 *
 * Copyright (c) 2014-2015 Charlike Mike Reagent, contributors.
 * Released under the MIT license.
 */

'use strict';

var pkg = require('./package.json');
var currentWeekNumber = require('./');
var argv = process.argv.slice(2);

function help() {
  console.log([
    '',
    '  ' + pkg.description,
    '',
    '  Example',
    '    $ current-week-number September 15, 2126',
    '    37',
    '    $ current-week-number August 07, 2016',
    '    32',
    '    $ current-week-number 02/17/2012',
    '    7',
    '    $ current-week-number',
    '    10'
  ].join('\n'));
}

if (argv.indexOf('--help') !== -1) {
  return help();
}

if (argv.indexOf('--version') !== -1) {
  console.log(pkg.version);
  return;
}

console.log(currentWeekNumber(argv.join('')));


index.js
/**
 * current-week-number <https://github.com/datetime/current-week-number>
 *
 * Copyright (c) 2014-2015 Charlike Mike Reagent, contributors.
 * Released under the MIT license.
 */

'use strict';

/**
 * Get week number of the current date/year or given valid `Date` string format
 *
 * **Example:**
 *
 * ```js
 * var currentWeekNumber = require('current-week-number');
 *
 * // june 27, 2014
 * currentWeekNumber();
 * //=> 26
 *
 * currentWeekNumber('March 24, 2015');
 * //=> 13
 *
 * currentWeekNumber(new Date('March 24, 2015'));
 * //=> 13
 *
 * currentWeekNumber('03/24/2016');
 * //=> 12, cuz' year is leap
 *
 * currentWeekNumber('August 07, 2015');
 * //=> 32
 *
 * currentWeekNumber(new Date('August 07, 2016'));
 * //=> 31
 *
 * currentWeekNumber('02/16/2015');
 * //=> 8
 * ```
 *
 * @name currentWeekNumber
 * @param  {String} `[date]` every valid Date-ish string format
 * @return {Number}
 * @api public
 */
module.exports = function currentWeekNumber(date) {
  var instance;

  if (typeof date === 'string' && date.length) {
    instance = new Date(date);
  } else if (date instanceof Date) {
    instance = date;
  } else {
    instance = new Date();
  }

  // Create a copy of this date object
  var target = new Date(instance.valueOf());

  // ISO week date weeks start on monday
  // so correct the day number
  var dayNr = (instance.getDay() + 6) % 7;

  // ISO 8601 states that week 1 is the week
  // with the first thursday of that year.
  // Set the target date to the thursday in the target week
  target.setDate(target.getDate() - dayNr + 3);

  // Store the millisecond value of the target date
  var firstThursday = target.valueOf();

  // Set the target to the first thursday of the year
  // First set the target to january first
  target.setMonth(0, 1);
  // Not a thursday? Correct the date to the next thursday
  if (target.getDay() !== 4) {
    target.setMonth(0, 1 + ((4 - target.getDay()) + 7) % 7);
  }

  // The weeknumber is the number of weeks between the
  // first thursday of the year and the thursday in the target week
  var weekNumber = 1 + Math.ceil((firstThursday - target) / 604800000);
  return weekNumber;
};
